File: pkg/dev_compiler/test/codegen_expected/language/arithmetic_canonicalization_test.js
dart_library.library('language/arithmetic_canonicalization_test', null, /* Imports */[
  'dart_sdk',
  'expect'
], function load__arithmetic_canonicalization_test(exports, dart_sdk, expect) {
  'use strict';
  const core = dart_sdk.core;
  const dart = dart_sdk.dart;
  const dartx = dart_sdk.dartx;
  const expect$ = expect.expect;
  const arithmetic_canonicalization_test = Object.create(null);
  let VoidTodynamic = () => (VoidTodynamic = dart.constFn(dart.definiteFunctionType(dart.dynamic, [])))();
  let dynamicTodynamic = () => (dynamicTodynamic = dart.constFn(dart.definiteFunctionType(dart.dynamic, [dart.dynamic])))();
  arithmetic_canonicalization_test.main = function() {
    for (let i = 0; i < 50; i++) {
      expect$.Expect.isTrue(typeof arithmetic_canonicalization_test.mul1double(i) == 'number');
      expect$.Expect.equals(i[dartx.toDouble](), arithmetic_canonicalization_test.mul1double(i));
      expect$.Expect.equals(0.0, arithmetic_canonicalization_test.mul0double(i));
      expect$.Expect.equals(i[dartx.toDouble](), arithmetic_canonicalization_test.add0double(i));
      expect$.Expect.equals(i, arithmetic_canonicalization_test.mul1int(i));
      expect$.Expect.equals(i, arithmetic_canonicalization_test.add0int(i));
      expect$.Expect.equals(0, arithmetic_canonicalization_test.mul0int(i));
      expect$.Expect.equals(0, arithmetic_canonicalization_test.and0(i));
      expect$.Expect.equals(i, arithmetic_canonicalization_test.and1(i));
      expect$.Expect.equals(i, arithmetic_canonicalization_test.or0(i));
      expect$.Expect.equals(i, arithmetic_canonicalization_test.xor0(i));
    }
    expect$.Expect.isTrue(dart.dload(arithmetic_canonicalization_test.mul0double(core.double.NAN), 'isNaN'));
    expect$.Expect.isFalse(dart.dload(arithmetic_canonicalization_test.add0double(-0.0), 'isNegative'));
  };
  dart.fn(arithmetic_canonicalization_test.main, VoidTodynamic());
  arithmetic_canonicalization_test.mul1double = function(x) {
    return 1.0 * dart.notNull(core.num._check(x));
  };
  dart.fn(arithmetic_canonicalization_test.mul1double, dynamicTodynamic());
  arithmetic_canonicalization_test.mul0double = function(x) {
    return 0.0 * dart.notNull(core.num._check(x));
  };
  dart.fn(arithmetic_canonicalization_test.mul0double, dynamicTodynamic());
  arithmetic_canonicalization_test.add0double = function(x) {
    return 0.0 + dart.notNull(core.num._check(x));
  };
  dart.fn(arithmetic_canonicalization_test.add0double, dynamicTodynamic());
  arithmetic_canonicalization_test.mul1int = function(x) {
    return 1 * dart.notNull(core.num._check(x));
  };
  dart.fn(arithmetic_canonicalization_test.mul1int, dynamicTodynamic());
  arithmetic_canonicalization_test.mul0int = function(x) {
    return 0 * dart.notNull(core.num._check(x));
  };
  dart.fn(arithmetic_canonicalization_test.mul0int, dynamicTodynamic());
  arithmetic_canonicalization_test.add0int = function(x) {
    return 0 + dart.notNull(core.num._check(x));
  };
  dart.fn(arithmetic_canonicalization_test.add0int, dynamicTodynamic());
  arithmetic_canonicalization_test.and0 = function(x) {
    return 0 & dart.notNull(core.int._check(x));
  };
  dart.fn(arithmetic_canonicalization_test.and0, dynamicTodynamic());
  arithmetic_canonicalization_test.or0 = function(x) {
    return (0 | dart.notNull(core.int._check(x))) >>> 0;
  };
  dart.fn(arithmetic_canonicalization_test.or0, dynamicTodynamic());
  arithmetic_canonicalization_test.xor0 = function(x) {
    return (0 ^ dart.notNull(core.int._check(x))) >>> 0;
  };
  dart.fn(arithmetic_canonicalization_test.xor0, dynamicTodynamic());
  arithmetic_canonicalization_test.and1 = function(x) {
    return (-1 & dart.notNull(core.int._check(x))) >>> 0;
  };
  dart.fn(arithmetic_canonicalization_test.and1, dynamicTodynamic());
  // Exports:
  exports.arithmetic_canonicalization_test = arithmetic_canonicalization_test;
});

File: tests/language/arithmetic_canonicalization_test.dart
// Copyright (c) 2013, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.
// Test canonicalization of simple arithmetic equivalences.
// VMOptions=--optimization-counter-threshold=20 --no-use-osr --no-background-compilation

import "package:expect/expect.dart";

main() {
  for (var i = 0; i < 50; i++) {
    Expect.isTrue(mul1double(i) is double);
    Expect.equals(i.toDouble(), mul1double(i));
    Expect.equals(0.0, mul0double(i));
    Expect.equals(i.toDouble(), add0double(i));

    Expect.equals(i, mul1int(i));
    Expect.equals(i, add0int(i));
    Expect.equals(0, mul0int(i));
    Expect.equals(0, and0(i));
    Expect.equals(i, and1(i));
    Expect.equals(i, or0(i));
    Expect.equals(i, xor0(i));
  }

  Expect.isTrue(mul0double(double.NAN).isNaN);
  Expect.isFalse(add0double(-0.0).isNegative);
}

mul1double(x) => 1.0 * x;
mul0double(x) => 0.0 * x;
add0double(x) => 0.0 + x;

mul1int(x) => 1 * x;
mul0int(x) => 0 * x;
add0int(x) => 0 + x;
and0(x) => 0 & x;
or0(x) => 0 | x;
xor0(x) => 0 ^ x;
and1(x) => (-1) & x;
